File: sort/quick_sort/java/quick_sort.java
//quick_sort classic algorithm

package quick_sort;

import java.util.Random;

public class Main {

    public static  void QuickSort(int[] Data, int Low, int High)
    {
        if(Low<High)
        {
            int Part = Partition(Data, Low, High);
            QuickSort(Data, Low, Part - 1);
            QuickSort(Data, Part + 1, High);
        }

    }

    public static int Partition(int[] Data, int Low, int High)
    {
        int Pivot = Data[High];
        int i=Low-1;
        for(int j=Low;j<High;j++)
        {
            if(Data[j]<=Pivot)
            {
                i++;
                int temp = Data[i];
                Data[i] = Data[j];
                Data[j] = temp;
            }
        }
        if(Data[High]<=Data[i+1])
        {
            int temp = Data[i+1];
            Data[i+1] = Data[High];
            Data[High] =temp;
        }
        return i+1;
    }

    public static int[] GenerateRandom(int ElementLength)
    {
        int[] Garbage = new int[ElementLength];
        Random Random = new Random();
        for(int i=0;i<Garbage.length;i++) Garbage[i] = Random.nextInt(ElementLength);
        return  Garbage;
    }

    public static void main(String[] args) {
        int[] test = new int[10];
        test = GenerateRandom(10);
        for (int i=0;i<10;i++) System.out.print(test[i]+",");
        System.out.println();
        //how to use? 1st param: array of int, 2nd param: starting index, 3rd param: ending index, if it the last one, substract it by one.
        QuickSort(test,0,test.length-1);
        for (int i=0;i<10;i++) System.out.print(test[i]+",");
    }
}

File: sort/quick_sort/kotlin/quicksort.kt
fun divideArr(array : ArrayList<Int>,low : Int, top : Int) : Int{
    val pivot = array[top]
    var i = (low - 1)
    for (j in low .. top){
        if(array[j] <= pivot){
            i++

            val t = array[i]
            array[i] = array[j]
            array[j] = t
        }
    }

    val t = array[i+1]
    array[i+1] = array[top]
    array[top] = t

    return i+1
}

fun quickSort(array : ArrayList<Int>, low : Int, top : Int){
    if(low < top){
        val p = divideArr(array,low,top)

        quickSort(array,low,p-1)
        quickSort(array,p+1,top)
    }
}

fun printArray(array : ArrayList<Int>){
    val n = array.size
    for ( i in 0 .. n){
        print(array[i].toString() + " ")
    }
    println("--------------")
}

fun main(args: Array<String>) {
    val array : ArrayList<Int> = ArrayList()
    array.apply {
        add(16)
        add(65)
        add(36)
        add(524)
        add(8431)
        add(3)
        add(1)
        add(6548)
        add(100)
        add(1656)
    }

    val n = array.size
    quickSort(array, 0, n-1)
    println("Array sorted by QuickSort")
    printArray(array)
}